metrics/opencensus_exporter.go

```diff
@@ -36,3 +55,82 @@ func newOpenCensusExporter(config *metricsConfig, logger *zap.SugaredLogger) (vi
 	view.RegisterExporter(e)
 	return e, nil
 }
+
+func defaultResourceDescriptor(ctx context.Context) (*resource.Resource, error) {
+	if r, ok := metricskey.FromContext(ctx); ok {
+		return r, nil
+	}
+	m := tag.FromContext(ctx)
+
+	if _, ok := m.Value(tag.MustNewKey(metricskey.LabelRevisionName)); ok {
+		return &resource.Resource{
+			Type:   metricskey.ResourceTypeKnativeRevision,
+			Labels: extractLabelsFromMap(m, metricskey.KnativeRevisionLabels.UnsortedList()...),
+		}, nil
+	}
+	if _, ok := m.Value(tag.MustNewKey(metricskey.LabelTriggerName)); ok {
+		return &resource.Resource{
+			Type:   metricskey.ResourceTypeKnativeTrigger,
+			Labels: extractLabelsFromMap(m, metricskey.KnativeTriggerLabels.UnsortedList()...),
+		}, nil
+	}
+	if _, ok := m.Value(tag.MustNewKey(metricskey.LabelEventSource)); ok {
+		return &resource.Resource{
+			Type:   metricskey.ResourceTypeKnativeSource,
+			Labels: extractLabelsFromMap(m, metricskey.KnativeSourceLabels.UnsortedList()...),
+		}, nil
+	}
+	if _, ok := m.Value(tag.MustNewKey(metricskey.LabelBrokerName)); ok {
+		return &resource.Resource{
+			Type:   metricskey.ResourceTypeKnativeBroker,
+			Labels: extractLabelsFromMap(m, metricskey.KnativeBrokerLabels.UnsortedList()...),
+		}, nil
+	}
+	return nil, nil
+}
+
+func extractLabelsFromMap(t *tag.Map, labels ...string) map[string]string {
+	ret := make(map[string]string, 0, len(labels))
+	for _, l := range labels {
+		k := tag.MustNewKey(l)
+		if v, ok := t.Value(k); ok {
+			ret[l] = v
+		} else {
+			ret[l] = "unknown"
+		}
+	}
+	return ret
+}
+
+// credentialFetcher attempts to locate a secret containing TLS credentials
+// for communicating with the OpenCensus Agent. To do this, it first looks
+// for a secret named "<component>-opencensus", then for a generic
+// "opencensus" secret.
+func credentialFetcher(component string, lister clientv1.SecretLister, logger *zap.SugaredLogger) credentials.TransportCredentials {
+	if lister == nil {
+		logger.Errorf("No secret lister provided for component %q; cannot use requireSecure=true", component)
+		return nil
+	}
+	return credentials.NewTLS(&tls.Config{
+		GetClientCertificate: func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+			// We ignore the CertificateRequestInfo for now, and hand back a single fixed certificate.
+			// TODO(evankanderson): maybe do something SPIFFE-ier?
+			cert, err := certificateFetcher(component+"-opencensus", lister)
+			if errors.IsNotFound(err) {
+				cert, err = certificateFetcher("opencensus", lister)
+			}
+			if err != nil {
+				return nil, fmt.Errorf("Unable to fetch opencensus secret for %q, cannot use requireSecure=true: %+v", component, err)
+			}
+			return &cert, err
+		},
+	})
+}
+
+func certificateFetcher(secretName string, lister clientv1.SecretLister) (tls.Certificate, error) {
+	secret, err := lister.Secrets(system.Namespace()).Get(secretName)
```

Comment:
This has a problem similar to https://github.com/knative/pkg/issues/842 where the exporter should be re-created when the Secret changes.

Reply:
This fetches the Secret (from the Informer cache) every time that a new TLS connection is constructed, because the grpc `credentials.TransportCredentials` constructed from `NewTLS` will use the `tls.Config` to start the handshake: https://github.com/grpc/grpc-go/blob/master/credentials/tls.go#L70

In turn, the `tls.Config` is constructed with a [`GetClientCertificate` function, which says](https://godoc.org/crypto/tls#Config):

> GetClientCertificate, if not nil, is called when a server requests a
> certificate from a client. If set, the contents of Certificates will
> be ignored.
>
> If GetClientCertificate returns an error, the handshake will be
> aborted and that error will be returned. Otherwise
> GetClientCertificate must return a non-nil Certificate. If
> Certificate.Certificate is empty then no certificate will be sent to
> the server. If this is unacceptable to the server then it may abort
> the handshake.
>
> GetClientCertificate may be called multiple times for the same
> connection if renegotiation occurs or if TLS 1.3 is in use.

This is why we need to feed the SecretLister down into the OpenCensus agent, so that it can fetch the secret whenever the library attempts to `grpc.Dial()` the server.

Reply:
Ah, I see, thanks for the detailed response! 

In terms of the keeping the SecretLister package local though, couldn't the metrics package maintain its own in-cluster kubeclient too?